File: 1901-find-a-peak-element-ii/1901-find-a-peak-element-ii.py
class Solution:

    def solve(self, mat, n, m, col):
        maxvalue = -1
        index = -1

        for i in range(n):
            if mat[i][col] > maxvalue:
                maxvalue = mat[i][col]
                index = i

        return index

    def findPeakGrid(self, mat):

        n = len(mat)
        m = len(mat[0])

        low = 0
        high = m - 1

        while low <= high:
            mid = (low + high) // 2

            mri = self.solve(mat, n, m, mid)

            left = mat[mri][mid - 1] if mid - 1 >= 0 else -1
            right = mat[mri][mid + 1] if mid + 1 < m else -1

            if mat[mri][mid] > left and mat[mri][mid] > right:
                return [mri, mid]

            elif mat[mri][mid] < left:
                high = mid - 1

            else:
                low = mid + 1

        return [-1, -1]

File: 1901-find-a-peak-element-ii/1901-find-a-peak-element-ii.java
class Solution {

    int solve(int[][] mat, int n, int m, int col) {
        int maxvalue = -1;
        int index = -1;

        for(int i = 0; i < n; i++) {
            if(mat[i][col] > maxvalue) {
                maxvalue = mat[i][col];
                index = i;
            }
        }

        return index;
    }

    public int[] findPeakGrid(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;

        int low = 0;
        int high = m - 1;

        while(low <= high) {
            int mid = (low + high) / 2;

            int mri = solve(mat, n, m, mid);

            int left = mid - 1 >= 0
                ? mat[mri][mid - 1] : -1;

            int right = mid + 1 < m
                ? mat[mri][mid + 1] : -1;

            if(mat[mri][mid] > left &&
               mat[mri][mid] > right) {
                return new int[]{mri, mid};
            }

            else if(mat[mri][mid] < left) {
                high = mid - 1;
            }

            else {
                low = mid + 1;
            }
        }

        return new int[]{-1, -1};
    }
}